### PR TITLE
authn_wasm: Add skeleton authn filter in WASM.

### PR DESCRIPTION
**Pull request by yangminzhu**
https://github.com/istio/istio/issues/15772

### CODE REVIEW

File: src/envoy/http/authn_wasm/config.cc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+using ::google::protobuf::TextFormat;
+using namespace google::protobuf::util;
+
+void PluginRootContext::onConfigure(std::unique_ptr<WasmData> configuration) {
+  logDebug("called PluginRootContext::onConfigure()");
+  JsonParseOptions json_options;
+  Status status = JsonStringToMessage(configuration->toString(),
+                                      &filter_config_, json_options);
+  if (status != Status::OK) {
```
Comment (incfly) [inline src/envoy/http/authn_wasm/config.cc]: if the status is not okay, will the filter stuck in a unsuable state? Is this err handling same in compile-in mode filter?

File: src/envoy/http/authn_wasm/config.cc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// imports from the low-level API
+using Common::Wasm::Null::NullVmPluginFactory;
+using Common::Wasm::Null::Plugin::logDebug;
+using Common::Wasm::Null::Plugin::logError;
+
+using ::google::protobuf::TextFormat;
+using namespace google::protobuf::util;
+
+void PluginRootContext::onConfigure(std::unique_ptr<WasmData> configuration) {
+  logDebug("called PluginRootContext::onConfigure()");
```
Comment (incfly) [inline src/envoy/http/authn_wasm/config.cc]: different filter will share the same function exported by host? maybe we should then add 'authn' prefix to simulate log scope?

File: src/envoy/http/authn_wasm/config.cc [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  return Http::FilterHeadersStatus::Continue;
+}
+
+// Registration glue
+
+Common::Wasm::Null::NullVmPluginRootRegistry* context_registry_{};
+
+class AuthnWasmFactory : public Common::Wasm::Null::NullVmPluginFactory {
+ public:
+  const std::string name() const override { return "envoy.wasm.authn_wasm"; }
```
Comment (incfly) [inline src/envoy/http/authn_wasm/config.cc]: authn_wasm seems redudant? wasm is in prefix already.

File: src/envoy/http/authn_wasm/config.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+using istio::envoy::config::filter::http::authn::v2alpha1::FilterConfig;
+using StringView = absl::string_view;
+using Common::Wasm::Null::NullVmPluginRootRegistry;
+using Common::Wasm::Null::Plugin::Context;
+using Common::Wasm::Null::Plugin::ContextFactory;
+using Common::Wasm::Null::Plugin::RootContext;
+using Common::Wasm::Null::Plugin::RootFactory;
+using Common::Wasm::Null::Plugin::WasmData;
+
+// PluginRootContext is the root context for all streams processed by the
```
Comment (incfly) [inline src/envoy/http/authn_wasm/config.h]: is this comment accurate? i saw the `wasm_filter.md` is documented this is a per vm instance.

File: src/envoy/http/authn_wasm/config.cc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+void PluginRootContext::onTick() {
+  logDebug("called PluginRootContext::onTick()");
+};
+
+void PluginContext::onCreate() { logDebug("called PluginContext::onCreate()"); }
+
+Http::FilterHeadersStatus PluginContext::onRequestHeaders() {
+  const auto& config = filter_config();
+  std::string out_str;
+  TextFormat::PrintToString(config, &out_str);
```
Comment (incfly) [inline src/envoy/http/authn_wasm/config.cc]: serialized computation is invoked for every request header as data path, might not added for now?

if WASM has some feature to load runtime feature (log level) and we can then put that computation there? same thing, i think logDebug is a function not a macro, and the DebugString is always invoked?

File: src/envoy/http/authn_wasm/config.cc (hunk 1 above)
Comment (yangminzhu) [reply src/envoy/http/authn_wasm/config.cc]: I think Envoy will just rejects the config, the filter will not be created.

File: src/envoy/http/authn_wasm/config.h (hunk 4 above)
Comment (yangminzhu) [reply src/envoy/http/authn_wasm/config.h]: I copied from other files, why is this not accurate? It talks about streams and thread but doesn't mention VMs?

File: src/envoy/http/authn_wasm/config.cc (hunk 2 above)
Comment (yangminzhu) [reply src/envoy/http/authn_wasm/config.cc]: That's a good point but I think it should be solved in the WASM layer, so I will just add a TODO at this moment.

File: src/envoy/http/authn_wasm/config.cc (hunk 5 above)
Comment (yangminzhu) [reply src/envoy/http/authn_wasm/config.cc]: I think this is another issue that should be resolved in the WASM level. I can just remove this for now.

**Comment by yangminzhu** [comment]
@incfly thanks for the review, comments addressed, PTAL.

File: src/envoy/http/authn_wasm/config.cc (hunk 1 above)
Comment (incfly) [reply src/envoy/http/authn_wasm/config.cc]: but how you let envoy know that? The status is not returned or consumed here.

File: src/envoy/http/authn_wasm/config.h (hunk 4 above)
Comment (incfly) [reply src/envoy/http/authn_wasm/config.h]: Yes.`wasm_filter.md` references the VM not the thread. @mandarjog and @jplevyak in last weekly meeting, we talked about the mapping of VM/worker threads, I can't remember the details though to confirm if it's 1:1 mapping. This seems closed to existing Envoy filter context, not sure how much applied to Wasm one.

> At bootstrap time, a root context is created. The root context
has the same lifetime as the VM/runtime instance and acts as a target
for any interactions which happen at initial setup. It is also used for
interactions that outlive a request.

File: src/envoy/http/authn_wasm/config.cc (hunk 2 above)
Comment (incfly) [reply src/envoy/http/authn_wasm/config.cc]: Have you added TODO here? didn't see the diff yet.

File: src/envoy/http/authn_wasm/config.cc (hunk 5 above)
Comment (incfly) [reply src/envoy/http/authn_wasm/config.cc]: sounds good. might worth file a feature request.

/cc @PiotrSikora @jplevyak

File: src/envoy/http/authn_wasm/config.cc (hunk 2 above)
Comment (incfly) [reply src/envoy/http/authn_wasm/config.cc]: never mind see it above.

**Pull request merged**